Add LightStep propagation support.

csabakos opened the pull request:
This adds support for LightStep propagation (`ot-tracer-traceid`, `ot-tracer-spanid`, `ot-tracer-sampled` headers), as well as a method for combined propagation which allows one to use multiple propagations.

The purpose is to allow people to more easily transition between LightStep headers and B3 headers by consuming/emitting both.

adriancole commented:
can you please add a direct link to the specification including the header names and values you are referring to? I'm aware of lightstep's `x-ot-span-context`, but not aware of another convention. plus it seemed like OpenTracing was moving towards OpenTelemetry which was going with the w3c headers?

csabakos commented:
The [OpenTracing specification](https://opentracing.io/specification/) does *not* specify what HTTP headers should be used.

However, the `ot-tracer-*` headers appear to be used in the wild. Some references:
- [lightstep/lightstep-tracer-objc](https://github.com/lightstep/lightstep-tracer-objc/blob/master/Pod/Classes/LSTracer.m#L146)
- [stripe/veneur](https://github.com/stripe/veneur/blob/master/trace/opentracing.go#L43)

> plus it seemed like OpenTracing was moving towards OpenTelemetry which was going with the w3c headers?

I believe this is correct, but existing OpenTracing users could benefit from this patch to provide a transition path between the `ot-tracer-*` headers and the `B3` headers, for example in case they are migrating away from OpenTracing to a Zipkin-based system).

adriancole commented:
this format is lightstep's format as noted also in the veneur code. let's
not brand this opentracing. if you want lightstep's format, maybe retitle
the issue etc.

On Thu, Jan 16, 2020 at 1:29 AM Csaba Kos <notifications@github.com> wrote:

> The OpenTracing specification <https://opentracing.io/specification/>
> does *not* specify what HTTP headers should be used.
>
> However, the ot-tracer-* headers appear to be used in the wild. Some
> references:
>
>    - lightstep/lightstep-tracer-objc
>    <https://github.com/lightstep/lightstep-tracer-objc/blob/master/Pod/Classes/LSTracer.m#L146>
>    - stripe/veneur
>    <https://github.com/stripe/veneur/blob/master/trace/opentracing.go#L43>
>
> plus it seemed like OpenTracing was moving towards OpenTelemetry which was
> going with the w3c headers?
>
> I believe this is correct, but existing OpenTracing users could benefit
> from this patch to provide a transition path between the ot-tracer-*
> headers and the B3 headers, for example in case they are migrating away
> from OpenTracing to a Zipkin-based system).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/openzipkin-contrib/brave-opentracing/pull/98?email_source=notifications&email_token=AAAPVV7OQKVXMJOD6E6YS23Q55BX5A5CNFSM4KG3HC7KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJBEDBY#issuecomment-574767495>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAPVVZ55APAGI2FRGY7DIDQ55BX5ANCNFSM4KG3HC7A>
> .
>

csabakos commented:
@adriancole Sorry, I didn't realize that the headers are vendor-specific. I updated the code accordingly and renamed the pull request, but now I wonder if `brave-opentracing` is the best place for this code or perhaps creating a vendor specific library would be better.

csabakos commented:
Thanks @adriancole, I will work with LightStep to find a better home for `LightStepPropagation.java`.

The other piece (composite propagation) seems generally useful (as evidenced by `propagation-stackdriver` and this pull request), so I've created https://github.com/openzipkin/brave/pull/1060 for that.
